streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -722,52 +729,58 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
         failure     <- Ref.make[Cause[OutErr1]](Cause.empty)
-        pull        <- (queueReader >>> self).toPullIn(scope)
+        pull        <- (queueReader >>> self).toPullInAlt(scope)
         _ <- pull
-               .foldCauseZIO(
-                 cause =>
-                   failure.update(_ && cause) *>
-                     outgoing.offer(Exit.fail(Left(()))) *>
-                     Exit.failCause(Cause.unit),
-                 {
-                   case Left(outDone) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *> outgoing.offer(Exit.fail(Right(outDone)))
-                   case Right(outElem) =>
-                     for {
-                       latch <- Promise.make[Nothing, Unit]
-                       _ <- permits.withPermit {
-                              latch.succeed(()) *>
-                                ZIO.uninterruptibleMask { restore =>
-                                  restore(errorSignal.await).raceFirstAwait(
-                                    restore(f(outElem))
-                                      .catchAllCause(cause => failure.update(_ && cause) *> Exit.failCause(Cause.unit))
-                                  )
-                                }.foldCauseZIO(
-                                  _ => outgoing.offer(Exit.fail(Left(()))) *> errorSignal.refailCause(Cause.unit),
-                                  elem => outgoing.offer(Exit.succeed(elem))
-                                )
-                            }.forkIn(scope)
-                       _ <- latch.await
-                     } yield ()
-                 }
+               .flatMap(outElem =>
+                 for {
+                   latch <- Promise.make[Nothing, Unit]
+                   _ <- permits
+                          .withPermit(
+                            latch.succeed(()) *> f(outElem)
+                              .foldCauseZIO(
+                                cause =>
+                                  failure.update(_ && cause).unless(cause.isInterrupted) *>
+                                    errorSignal.succeed(()) *>
+                                    outgoing.offer(ZChannel.failLeftUnit),
+                                elem => outgoing.offer(Exit.succeed(elem))
+                              )
+                          )
+                          .fork
+                   _ <- latch.await
+                 } yield ()
                )
                .forever
-               .interruptible
+               .catchAllCause(cause =>
+                 cause.failureOrCause match {
+                   case Left(x: Left[OutErr, OutDone]) =>
+                     failure.update(_ && Cause.fail(x.value)) *>
+                       outgoing.offer(ZChannel.failLeftUnit) *>
+                       ZChannel.failUnit
+                   case Left(x: Right[OutErr, OutDone]) =>
+                     permits.withPermits(n.toLong)(ZIO.unit) *>
```

Comment:
Previously this was `interruptible`. Is it still needed?

Reply:
Yeah the acquire is uninterruptible, thank you